### PR TITLE
add an option to disable Rpc-Service response header for tchannel transport

### PR DESCRIPTION
**Pull request by kishoresenji**
Adding an option to disable adding $rpc$-service header for tchannel transport. The config (excludeServiceHeaderInResponse) is a TransportOption and by default it is false and preserves existing functionality. One would have to explicitly turn it on by using ExcludeServiceHeaderInResponse() transport option. If excludeServiceHeaderInResponse is set to true, the $rpc$-service header will not be added to response.

The $rpc$-service header does not play well if there are tchannel-go [inbound](https://github.com/yarpc/yarpc-go/blob/dev/transport/tchannel/header.go#L242-L243) servers in the call graph. The service name validation makes it very difficult to upgrade to yarpc.

### CODE REVIEW

**Comment by rabbbit** [comment]
I think we probably won't do this, but for completeness - could you expand the PR description on how long we expect this header to be present? How and if do we expect to get rid of it?

Some (not convincing) ideas:
- we could add a stat for this
- we could add a "timeout" for this - for example, it would gradually stop working over a month
- we could make it return the header for 1% of the calls so that (hopefully) someone catches any issues

I am not a big fan of any of the ideas, but I would like this to be included in the PR message - just for the future to know what (the hell) we were thinking when adding this ;)

**Comment by kishoresenji** [comment]
> I think we probably won't do this, but for completeness - could you expand the PR description on how long we expect this header disabling behaviour to be present? How and if do we expect to get rid of it?
> 
> Some (not convincing) ideas:
> 
> * we could add a stat for this
> * we could add a "timeout" for this - for example, it would gradually stop working over a month
> * we could make it return the header for 1% of the calls so that (hopefully) someone catches any issues
> 
> I am not a big fan of any of the ideas, but I would like this to be included in the PR message - just for the future to know what (the hell) we were thinking when adding this ;)

The api is public and I do not know the best practices followed by yarpc for deprecating features, but it will be similar to opensource best practices. This can be seen as a feature too, ie yarpc does not have to force security/visibility for everyone. One can choose whether they want the extra security/visibility on routing errors or not. 
Personally I would not add 1% failures or timer or anything like that. I will defer that to this repo maintainers/committers.

**Comment by gopalchouhan** [comment]
In my opinion, yarpc is doing the right thing and this validation was added for something meaningful.  I also believe that it is not the right way to fix this. If we do this, removing the service validation may become a norm for doing migration from tchannel-go to yarpc-go to avoid any issues and we might be back to square one with the same issue it was intended to fix.

**Comment by kishoresenji** [comment]
> I believe this validation was added for something meaningful. I also believe that it may not be the right way to fix this. If we do this, removing the service validation may become a norm for doing migration from tchannel-go to yarpc-go to avoid any issues and we might be back to square one with the same issue it was intended to fix.

Let us internally sync on this. We are not removing validation. We are showing a path for easy upgrade. I have a similar PR on tchannel as well and pursuing both options.

File: transport/tchannel/options.go
Comment (AllenLuUber) [inline transport/tchannel/options.go:226]: With is not 100% necessary here. `DisableRpcServiceHeader()` is a better name

**Review by AllenLuUber — APPROVED** [review]
Just a naming suggestion. Otherwise LGTM

File: transport/tchannel/options.go
Comment (gopalchouhan) [inline transport/tchannel/options.go:64]: It is not clear if it disables the serviceHeader in the response. Can we make it more explicit like excludeServiceHeaderInResponse or something related?

File: transport/tchannel/options.go
Comment (witriew) [inline transport/tchannel/options.go]: Using a negative setter for bools is usually a bad pattern.  Can we invert this and document the default value if not leveraged?

This way, `option.disableRpcServiceHeader = false` avoids a double negative.

For example,
```
func AddRpcServiceHeader(val bool) TransportOption
```
sidenote: `Add...`, `Use...` or `Enable...` or any equivalent such naming.

File: transport/tchannel/options.go
Comment (kishoresenji) [reply transport/tchannel/options.go]: Done. thanks

File: transport/tchannel/options.go
Comment (kishoresenji) [reply transport/tchannel/options.go]: Using excludeServiceHeaderInResponse as suggested. thanks

**Review by gopalchouhan — APPROVED** [review]
LGTM

File: transport/tchannel/options.go
Comment (kishoresenji) [reply transport/tchannel/options.go]: The reason I chose this way is so that I can use the zero value to effectively keep the existing functionality easily. But if we prefer I can make the change and carefully ensure we set the value to true where ever it is instantiated.

Comment (gopalchouhan) [reply transport/tchannel/options.go]: IMO, Having exclude prefix more suitable in this case due to its opt-out nature. We want the services to explicitly pass this exclude option to disable the service header in the response.

Comment (kishoresenji) [reply transport/tchannel/options.go]: What @witriew is referring to is the internal data structure. External api will still be exclude. Internally we can still have struct field as `addServiceHeaderInResponse`. We just need to ensure that field is set to true by default where Transport is created.

Comment (kishoresenji) [reply transport/tchannel/options.go]: oh Sorry @witriew  is referring to the api. Yea, the api should be Exclude. Internally we can choose the struct to have a field to addServiceHeaderInResponse set to true.

Comment (witriew) [reply transport/tchannel/options.go]: I'm referring to both the API and the internal param.

> IMO, Having exclude prefix more suitable in this case due to its opt-out nature. We want the services to explicitly pass this exclude option to disable the service header in the response.

The default internally can be whatever we choose it to be, as long as it's documented what the behavior is.  The API provided doesn't necessarily need to expose what the internal setting was, but rather, its simplicity when it's being used should be considered, as in, setting a `disable` to `false` feels awkward.

Comment (kishoresenji) [reply transport/tchannel/options.go]: @witriew  - just to conclude this - can you be explicit in what you are suggesting? API and the internal field please.

Comment (kishoresenji) [reply transport/tchannel/options.go]: Btw, please note that we are not setting `option.disableRpcServiceHeader = false` instead we are doing this `option.disableRpcServiceHeader = true`

Comment (AllenLuUber) [reply transport/tchannel/options.go]: yea we should be not setting the value to `false`. The zero value is already `false` and we should utilize that.

File: transport/tchannel/handler_test.go
Comment (AllenLuUber) [inline transport/tchannel/handler_test.go:812]: nit: if all are ignored with _, you don't need to specify:
`func( inboundCallResponse,  tchannel.Format,  headerCase) responseWriter`

Comment (kishoresenji) [reply transport/tchannel/handler_test.go]: updated. thanks

File: transport/tchannel/options.go
Comment (witriew) [reply transport/tchannel/options.go]: > @witriew - just to conclude this - can you be explicit in what you are suggesting? API and the internal field please.

Yah - well, the internal field should match what the API is setting and mentioned briefly [above](https://github.com/yarpc/yarpc-go/pull/2120#discussion_r797207606):
```
func AddRpcServiceHeader(val bool) TransportOption
```
taking in the value to set it to, but defaulting to `true` internally.

But if @AllenLuUber is fine with this, it's his call.

Comment (AllenLuUber) [reply transport/tchannel/options.go]: yea `DisableRpcServiceHeader ` sounds good enough.

**Pull request merged**
